File: OpenAIJsonWrapper/OpenAIJsonWrapper/wrapper.py
import base64
import json
import re
from pathlib import Path
from typing import Any, List, Optional, Tuple, Union, Dict

TOOL_MARKER_START = "```json"
TOOL_MARKER_END = "```"

class OpenAIJsonWrapper:
    """
    OpenAI 客户端封装，用于强制模型输出 JSON 并自动解析。
    用户只需提供数据结构定义（dict/list），框架会自动拼接 System Prompt 并解析返回。
    """

    def __init__(
        self, 
        client: Any, 
        model: str = "gpt-3.5-turbo", 
        target_structure: Optional[Any] = None,
        requirements: Optional[Union[str, List[str]]] = None,
        background: Optional[str] = None
    ):
        """
        :param client: OpenAI 风格的客户端对象 (需具备 chat.completions.create 方法)
        :param model: 默认使用的模型名称
        :param target_structure: 默认的输出结构定义
        :param requirements: 默认的特定需求说明
        :param background: 默认的背景信息
        """
        self.client = client
        self.model = model
        self.target_structure = target_structure
        self.requirements = requirements
        self.background = background

    def _build_system_prompt(self, target_structure: Any, requirements: Optional[Union[str, List[str]]] = None, background: Optional[str] = None) -> str:
        structure_str = json.dumps(target_structure, indent=2, ensure_ascii=False)
        prompt = (
            "You are a helpful assistant that MUST output your response in a specific JSON format.\n"
        )
        
        if background:
            prompt += f"\nBackground Information:\n{background}\n"
            
        prompt += f"\nThe required JSON structure is:\n{structure_str}\n\n"
        
        if requirements:
            if isinstance(requirements, list):
                req_text = "\n".join([f"- {r}" for r in requirements])
            else:
                req_text = requirements
            prompt += f"Specific Requirements:\n{req_text}\n\n"
            
        prompt += (
            "Rules:\n"
            f"1. Your final JSON data MUST be wrapped between '{TOOL_MARKER_START}' and '{TOOL_MARKER_END}' markdown blocks.\n"
            "2. Everything before the code block is considered your reasoning or conversational text.\n"
            "3. Ensure the JSON inside the block is valid and matches the requested structure strictly.\n"
        )
        return prompt

    def _to_list(self, x: Optional[Union[str, List[str]]]) -> List[str]:
        if x is None:
            return []
        return x if isinstance(x, list) else [x]

    def _parse_content(self, text: str) -> Tuple[str, Any, Optional[str]]:
        """
        解析模型文本中内嵌的 JSON 块。
        返回 (reasoning_content, parsed_data, error_message)
        """
        if not text:
            return "", None, "Empty content"

        # 处理思考链：如果存在 </think>，则只保留其之后的内容
        think_end_marker = "</think>"
        if think_end_marker in text:
            # 找到最后一个 </think> 的位置并截取其后的内容
            parts = text.split(think_end_marker)
            text = parts[-1].strip()

        # 1) 查找特定的 Markdown JSON 块
        # 使用正则匹配 ```json ... ```，尽量抓取内部内容
        md_pattern = r"```json\s*([\s\S]*?)\s*```"
        match = re.search(md_pattern, text)
        if match:
            inner = match.group(1).strip()
            reasoning = text[:match.start()].strip()
            try:
                parsed = json.loads(inner)
                return reasoning, parsed, None
            except Exception as e:
                # 如果标准加载失败，尝试简单的清理（比如去掉可能的尾随逗号）再解析
                try:
                    # 极简修复：去除非法字符
                    cleaned = re.sub(r',\s*([\]}])', r'\1', inner)
                    parsed = json.loads(cleaned)
                    return reasoning, parsed, None
                except:
                    return text, None, f"JSON parse error in markdown block: {e}"

        # 2) 备选：尝试从末尾提取最后一个 JSON 结构 (兼容没有标记的情况)
        for opener, closer in (("[", "]"), ("{", "}")):
            idx = text.rfind(opener)
            if idx != -1:
                cand = text[idx:].strip()
                # 寻找匹配的闭合符号位置
                last_closer = cand.rfind(closer)
                if last_closer != -1:
                    cand = cand[:last_closer+1]
                try:
                    parsed = json.loads(cand)
                    reasoning = text[:idx].strip()
                    return reasoning, parsed, None
                except Exception:
                    continue

        return text, None, "No JSON structure found"

    def _encode_image(self, image_source: str) -> str:
        """将图片路径或 URL 转为 base64 data URI 或直接 URL。"""
        if image_source.startswith("http://") or image_source.startswith("https://"):
            return image_source
        path = Path(image_source)
        if not path.exists():
            raise FileNotFoundError(f"Image not found: {image_source}")
        ext = path.suffix.lower()
        mime_map = {
            ".jpg": "image/jpeg",
            ".jpeg": "image/jpeg",
            ".png": "image/png",
            ".gif": "image/gif",
            ".webp": "image/webp",
            ".bmp": "image/bmp",
        }
        mime = mime_map.get(ext, "image/jpeg")
        with open(path, "rb") as f:
            data = base64.b64encode(f.read()).decode("utf-8")
        return f"data:{mime};base64,{data}"

    def _normalize_content_part(self, part: Any) -> Any:
        """
        将用户传入的消息 part 标准化为 OpenAI 支持的格式。
        支持以下自定义类型:
          - {"type": "image_path", "image_path": "..."} -> {"type": "image_url", "image_url": {"url": ...}}
          - {"type": "image_url", "image_url": {...}} 原样
          - {"type": "text", "text": "..."} 原样
        """
        if not isinstance(part, dict):
            return part
        ptype = part.get("type")
        if ptype == "image_path":
            image_path = part.get("image_path") or part.get("path")
            if not image_path:
                raise ValueError("image_path part must have 'image_path' or 'path' field")
            url = self._encode_image(image_path)
            return {"type": "image_url", "image_url": {"url": url}}
        if ptype == "image_url":
            url = part.get("image_url")
            if isinstance(url, dict) and "url" in url:
                url_value = url["url"]
                if not (url_value.startswith("http://") or url_value.startswith("https://") or url_value.startswith("data:")):
                    url_value = self._encode_image(url_value)
                return {"type": "image_url", "image_url": {**url, "url": url_value}}
            if isinstance(url, str):
                if not (url.startswith("http://") or url.startswith("https://") or url.startswith("data:")):
                    url = self._encode_image(url)
                return {"type": "image_url", "image_url": {"url": url}}
            return part
        return part

    def _normalize_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
        """
        标准化单条消息:
        - role 不变
        - 如果 content 是字符串，原样保留
        - 如果 content 是 list，逐项进行 part 标准化
        """
        role = message.get("role")
        content = message.get("content")
        if isinstance(content, list):
            new_content = [self._normalize_content_part(p) for p in content]
            return {**message, "role": role, "content": new_content}
        return message

    def _normalize_messages(self, messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        return [self._normalize_message(m) for m in messages]

    def chat(
        self, 
        messages: List[Dict[str, str]], 
        target_structure: Optional[Any] = None, 
        requirements: Optional[Union[str, List[str]]] = None,
        extra_requirements: Optional[Union[str, List[str]]] = None,
        background: Optional[str] = None,
        model: Optional[str] = None,
        **kwargs
    ) -> Dict[str, Any]:
        """
        执行对话并自动解析结果。
        
        :param messages: 对话上下文。content 可以是字符串，也可以是 list（多模态），
                         list 中的 part 支持:
                         - {"type": "text", "text": "..."}
                         - {"type": "image_url", "image_url": {"url": "http(s)://..." | "data:..."} | "path/to/img.jpg"}
                         - {"type": "image_path", "image_path": "path/to/img.jpg"}  (本地路径或 URL)
        :param target_structure: 希望获取的 JSON 模板/定义（若为 None 则使用实例初始化时的值）
        :param requirements: 特定需求说明 (str 或 list)
        :param extra_requirements: 额外补充需求 (str 或 list)，会与 requirements 合并
        :param background: 背景背景知识/上下文 (str)
        :param model: 模型名称（若为 None 则使用实例初始化时的值）
        :return: 包含 'reasoning', 'data', 'raw', 'error' 的字典
        """
        # 优先级：方法参数 > 初始化参数
        target = target_structure if target_structure is not None else self.target_structure
        reqs = requirements if requirements is not None else self.requirements
        
        # 将基础需求与当前调用的额外需求合并
        combined_list = self._to_list(reqs) + self._to_list(extra_requirements)
        reqs_for_prompt = combined_list if combined_list else None
        bg = background if background is not None else self.background
        selected_model = model if model is not None else self.model

        if target is None:
            raise ValueError("target_structure must be provided either in __init__ or in chat()")

        # 复制消息列表并注入系统提示
        prompt = self._build_system_prompt(target, requirements=reqs_for_prompt, background=bg)
        
        new_messages = []
        has_system = False
        for m in messages:
            if m.get("role") == "system":
                sys_content = f"{prompt}\n\n{m['content']}"
                new_messages.append({"role": "system", "content": sys_content})
                has_system = True
            else:
                new_messages.append(self._normalize_message(m))
        
        if not has_system:
            new_messages.insert(0, {"role": "system", "content": prompt})

        # 调用模型
        response = self.client.chat.completions.create(
            model=selected_model,
            messages=new_messages,
            **kwargs
        )

        # 获取 content
        choice = response.choices[0]
        content = choice.message.content or ""
        
        reasoning, data, error = self._parse_content(content)
        
        return {
            "reasoning": reasoning,
            "data": data,
            "error": error,
            "raw_content": content,
            "response_id": getattr(response, "id", None)
        }

File: OpenAIJsonWrapper/go/src/wrapper.go
package openaijsonwrapper

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type OpenAIJsonWrapper struct {
	client           OpenAIClient
	model            string
	targetStructure  any
	requirements      []string
	background       *string
}

type ChatOptions struct {
	TargetStructure   any
	Requirements      string
	ExtraRequirements string
	Background        string
	Model             string
}

func New(
	client OpenAIClient,
	model string,
	targetStructure any,
	requirements []string,
	background string,
) *OpenAIJsonWrapper {
	return &OpenAIJsonWrapper{
		client:          client,
		model:           model,
		targetStructure: targetStructure,
		requirements:    requirements,
	}
}

func toList(s *string) []string {
	if s == nil {
		return nil
	}
	return []string{*s}
}

func (w *OpenAIJsonWrapper) buildSystemPrompt(targetStructure any, requirements []string, background *string) string {
	structureStr, _ := json.MarshalIndent(targetStructure, "", "  ")
	prompt := "You are a helpful assistant that MUST output your response in a specific JSON format.\n"

	if background != nil {
		prompt += fmt.Sprintf("\nBackground Information:\n%s\n", *background)
	}

	prompt += fmt.Sprintf("\nThe required JSON structure is:\n%s\n\n", string(structureStr))

	if len(requirements) > 0 {
		reqText := ""
		for _, r := range requirements {
			reqText += fmt.Sprintf("- %s\n", r)
		}
		prompt += fmt.Sprintf("Specific Requirements:\n%s\n", reqText)
	}

	prompt += fmt.Sprintf(
		"Rules:\n"+
			"1. Your final JSON data MUST be wrapped between '%s' and '%s' markdown blocks.\n"+
			"2. Everything before the code block is considered your reasoning or conversational text.\n"+
			"3. Ensure the JSON inside the block is valid and matches the requested structure strictly.\n",
		ToolMarkerStart, ToolMarkerEnd,
	)

	return prompt
}

func parseContent(text string) (string, json.RawMessage, *string) {
	if text == "" {
		return "", nil, strPtr("Empty content")
	}

	thinkEndMarker := "</think>"
	if idx := strings.Index(text, thinkEndMarker); idx != -1 {
		text = strings.TrimSpace(text[idx+len(thinkEndMarker):])
	}

	mdPattern := regexp.MustCompile("```json\\s*([\\s\\S]*?)\\s*```")
	if matches := mdPattern.FindStringSubmatch(text); len(matches) > 1 {
		inner := strings.TrimSpace(matches[1])
		reasoning := text[:strings.Index(text, matches[0])]

		if parsed, err := parseJSON(inner); err == nil {
			return reasoning, parsed, nil
		}

		cleaned := regexp.MustCompile(",(\\s*[\\]}])").ReplaceAllString(inner, "$1")
		if parsed, err := parseJSON(cleaned); err == nil {
			return reasoning, parsed, nil
		}

		errMsg := fmt.Sprintf("JSON parse error in markdown block")
		return reasoning, nil, &errMsg
	}

	for _, pair := range [][]string{{"[", "]"}, {"{", "}"}} {
		opener, closer := pair[0], pair[1]
		if idx := strings.LastIndex(text, opener); idx != -1 {
			cand := strings.TrimSpace(text[idx:])
			if lastCloser := strings.LastIndex(cand, closer); lastCloser != -1 {
				cand = cand[:lastCloser+1]
				if parsed, err := parseJSON(cand); err == nil {
					reasoning := strings.TrimSpace(text[:idx])
					return reasoning, parsed, nil
				}
			}
		}
	}

	return text, nil, strPtr("No JSON structure found")
}

func parseJSON(s string) (json.RawMessage, error) {
	var v any
	if err := json.Unmarshal([]byte(s), &v); err != nil {
		return nil, err
	}
	return json.Marshal(v)
}

func (w *OpenAIJsonWrapper) encodeImage(imageSource string) (string, error) {
	if strings.HasPrefix(imageSource, "http://") || strings.HasPrefix(imageSource, "https://") {
		return imageSource, nil
	}

	path := filepath.Clean(imageSource)
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return "", fmt.Errorf("image not found: %s", path)
	}

	ext := strings.ToLower(filepath.Ext(path))
	mime := map[string]string{
		".jpg":  "image/jpeg",
		".jpeg": "image/jpeg",
		".png":  "image/png",
		".gif":  "image/gif",
		".webp": "image/webp",
		".bmp":  "image/bmp",
	}[ext]
	if mime == "" {
		mime = "image/jpeg"
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("failed to read image: %w", err)
	}

	encoded := encodeBase64(data)
	return fmt.Sprintf("data:%s;base64,%s", mime, encoded), nil
}

func encodeBase64(data []byte) string {
	const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
	result := ""
	for i := 0; i < len(data); i += 3 {
		var n uint32
		remaining := len(data) - i
		n |= uint32(data[i]) << 16
		if remaining > 1 {
			n |= uint32(data[i+1]) << 8
		}
		if remaining > 2 {
			n |= uint32(data[i+2])
		}

		result += string(alphabet[n>>18&63]) + string(alphabet[n>>12&63])
		if remaining > 1 {
			result += string(alphabet[n>>6&63])
		} else {
			result += "="
		}
		if remaining > 2 {
			result += string(alphabet[n&63])
		} else {
			result += "="
		}
	}
	return result
}

func (w *OpenAIJsonWrapper) normalizeContentPart(part ContentPart) (ContentPart, error) {
	switch part.Type {
	case ContentPartTypeImagePath:
		url, err := w.encodeImage(part.ImagePath)
		if err != nil {
			return part, err
		}
		return ContentPart{
			Type:     ContentPartTypeImageURL,
			ImageURL: ImageURLValue{URL: url},
		}, nil

	case ContentPartTypeImageURL:
		switch v := part.ImageURL.(type) {
		case string:
			if strings.HasPrefix(v, "http://") || strings.HasPrefix(v, "https://") || strings.HasPrefix(v, "data:") {
				return ContentPart{
					Type:     ContentPartTypeImageURL,
					ImageURL: ImageURLValue{URL: v},
				}, nil
			}
			url, err := w.encodeImage(v)
			if err != nil {
				return part, err
			}
			return ContentPart{
				Type:     ContentPartTypeImageURL,
				ImageURL: ImageURLValue{URL: url},
			}, nil
		case ImageURLValue:
			if strings.HasPrefix(v.URL, "http://") || strings.HasPrefix(v.URL, "https://") || strings.HasPrefix(v.URL, "data:") {
				return part, nil
			}
			url, err := w.encodeImage(v.URL)
			if err != nil {
				return part, err
			}
			return ContentPart{
				Type:     ContentPartTypeImageURL,
				ImageURL: ImageURLValue{URL: url, Detail: v.Detail},
			}, nil
		}
	}
	return part, nil
}

func (w *OpenAIJsonWrapper) normalizeMessage(message Message) (Message, error) {
	if message.Content.Array != nil {
		normalized := make([]ContentPart, 0, len(message.Content.Array))
		for _, part := range message.Content.Array {
			np, err := w.normalizeContentPart(part)
			if err != nil {
				return message, err
			}
			normalized = append(normalized, np)
		}
		message.Content.Array = normalized
	}
	return message, nil
}

func (w *OpenAIJsonWrapper) Chat(messages []Message, options ChatOptions) (*ChatResult, error) {
	target := w.targetStructure
	if options.TargetStructure != nil {
		target = options.TargetStructure
	}
	if target == nil {
		return nil, fmt.Errorf("targetStructure must be provided")
	}

	reqs := w.requirements
	if options.Requirements != "" {
		reqs = append(reqs, options.Requirements)
	}
	if options.ExtraRequirements != "" {
		reqs = append(reqs, options.ExtraRequirements)
	}

	bg := w.background
	if options.Background != "" {
		bg = &options.Background
	}

	model := w.model
	if options.Model != "" {
		model = options.Model
	}

	systemPrompt := w.buildSystemPrompt(target, reqs, bg)

	newMessages := make([]Message, 0, len(messages)+1)
	hasSystem := false

	for _, m := range messages {
		if m.Role == "system" {
			var sysContent string
			if m.Content.String != nil {
				sysContent = systemPrompt + "\n\n" + *m.Content.String
			} else {
				arrContent, _ := json.Marshal(m.Content.Array)
				sysContent = systemPrompt + "\n\n" + string(arrContent)
			}
			newMessages = append(newMessages, Message{
				Role:    "system",
				Content: MessageContent{String: &sysContent},
			})
			hasSystem = true
		} else {
			normalized, err := w.normalizeMessage(m)
			if err != nil {
				return nil, err
			}
			newMessages = append(newMessages, normalized)
		}
	}

	if !hasSystem {
		newMessages = append([]Message{{
			Role:    "system",
			Content: MessageContent{String: &systemPrompt},
		}}, newMessages...)
	}

	request := ChatCompletionRequest{
		Model:    model,
		Messages: newMessages,
	}

	response, err := w.client.ChatCompletionsCreate(request)
	if err != nil {
		errStr := err.Error()
		return &ChatResult{
			Error:      &errStr,
			RawContent: "",
		}, nil
	}

	choice := response.Choices[0]
	content := ""
	if choice.Message.Content != nil {
		content = *choice.Message.Content
	}

	reasoning, data, parseErr := parseContent(content)

	return &ChatResult{
		Reasoning:  reasoning,
		Data:       data,
		Error:      parseErr,
		RawContent: content,
		ResponseID: &response.ID,
	}, nil
}

func strPtr(s string) *string {
	return &s
}
